daq/topology.py

```diff
@@ -417,24 +421,27 @@ def _add_dhcp_reflectors(self, acl_list):
         self._add_acl_rule(acl_list, dl_type='0x800', allow=False,
                            nw_proto=17, udp_src=67, udp_dst=68)
 
+    def _add_dot1x_incoming_rule(self, incoming_acl, secondary_acl):
+        for devices in self._set_devices.values():
+            for device in devices:
+                if device and device.gateway:
+                    self._add_acl_rule(incoming_acl,
+                                       vlan_vid=self._get_port_vlan(device.port.port_no),
+                                       eth_type=self._DOT1X_ETH_TYPE,
+                                       ports=device.gateway.get_possible_test_ports())
+                    self._add_acl_rule(secondary_acl,
```

Comment:
The key to factoring out a function or bit of code is to identify repetition and reduce it as much as possible. I see the same snippet of code 4x times right now, with only minor variances -- I think it would be a fair bit cleaner to consolidate those into one function that's called 4x times.  Something like:

self._add_dot1x_rule(incoming_acl, [device.port.port_no], vlan_vid=port_vlan)

Conceptually, the consolidates the concept "the rule that handles dot1x eth" into one place... and makes it much more clear/obvious that any change needs to hit all four places (otherwise somebody needs to remember to search the code for any other uses of DOT1X_ETH_TYPE, etc...

Reply:
Refactored _add_acl_rule call into a separate method